File: ios/RoutineChart/RoutineChart/Features/Child/TodayView/ChildTodayViewModel.swift
//
//  ChildTodayViewModel.swift
//  RoutineChart
//
//  ViewModel for child's today view
//

import Foundation
import Combine
import OSLog

@MainActor
final class ChildTodayViewModel: ObservableObject {
    @Published var children: [ChildProfile] = []
    @Published var selectedChild: ChildProfile?
    @Published var routines: [RoutineWithSteps] = []
    @Published var isLoading = false
    @Published var error: String?
    @Published var showSettings = false
    
    let dependencies: AppDependencies
    
    private let childRepo: ChildProfileRepository
    private let routineRepo: RoutineRepository
    private let stepRepo: RoutineStepRepository
    private let assignmentRepo: RoutineAssignmentRepository
    private let eventRepo: CompletionEventRepository
    private let completeStep: CompleteStepUseCase
    private let undoStep: UndoStepUseCase
    private let deriveStepCompletion: DeriveStepCompletionUseCase
    private let authRepository: AuthRepository
    private let userRepository: UserRepository
    
    private var familyId: String?
    private let today: String
    
    init(dependencies: AppDependencies) {
        self.dependencies = dependencies
        self.childRepo = dependencies.childRepo
        self.routineRepo = dependencies.routineRepo
        self.stepRepo = dependencies.stepRepo
        self.assignmentRepo = dependencies.assignmentRepo
        self.eventRepo = dependencies.eventRepo
        self.completeStep = dependencies.completeStep
        self.undoStep = dependencies.undoStep
        self.deriveStepCompletion = dependencies.deriveStepCompletion
        self.authRepository = dependencies.authRepository
        self.userRepository = dependencies.userRepository
        self.today = Date().localDayKey()
    }
    
    func loadData() async {
        isLoading = true
        defer { isLoading = false }
        
        do {
            // Get current authenticated user
            guard let authUser = authRepository.currentUser else {
                error = "Not signed in"
                return
            }
            
            // Get user record to find their familyId
            guard let user = try await userRepository.get(id: authUser.id) else {
                error = "User not found. Please join a family first."
                return
            }
            
            familyId = user.familyId
            
            // Load children for THIS user's family
            children = try await childRepo.getAll(familyId: user.familyId)
            
            // Auto-select first child
            if selectedChild == nil, let first = children.first {
                selectedChild = first
                await loadRoutines()
            }
        } catch {
            self.error = error.localizedDescription
            AppLogger.error("Failed to load data: \(error.localizedDescription)")
        }
    }
    
    func selectChild(_ child: ChildProfile) async {
        selectedChild = child
        await loadRoutines()
    }
    
    func toggleStep(routine: RoutineWithSteps, step: StepWithCompletion) async {
        guard let child = selectedChild, let familyId = familyId else { return }
        
        do {
            if step.isComplete {
                try await undoStep.execute(
                    familyId: familyId,
                    childId: child.id,
                    routineId: routine.routine.id,
                    stepId: step.step.id
                )
            } else {
                try await completeStep.execute(
                    familyId: familyId,
                    childId: child.id,
                    routineId: routine.routine.id,
                    stepId: step.step.id
                )
            }
            
            // Reload to show updated state
            await loadRoutines()
        } catch {
            self.error = error.localizedDescription
            AppLogger.error("Failed to toggle step: \(error.localizedDescription)")
        }
    }
    
    private func loadRoutines() async {
        guard let child = selectedChild, let familyId = familyId else { return }
        
        do {
            // CRITICAL: Pull routines from Firestore BEFORE loading from local
            // This ensures we have the latest data when the UI loads
            if let compositeRepo = routineRepo as? CompositeRoutineRepository,
               let authUser = authRepository.currentUser,
               let user = try? await userRepository.get(id: authUser.id) {
                do {
                    // First, upload any unsynced local changes
                    let uploaded = try await compositeRepo.uploadUnsynced(userId: user.id, familyId: familyId)
                    if uploaded > 0 {
                        AppLogger.ui.info("✅ Uploaded \(uploaded) unsynced routine(s) before loading child view")
                    }
                    
                    // Then, pull any remote changes (this will also pull steps)
                    let pulled = try await compositeRepo.pullRoutines(userId: user.id, familyId: familyId)
                    if pulled > 0 {
                        AppLogger.ui.info("✅ Pulled \(pulled) routine(s) from Firestore before loading child view")
                    }
                } catch {
                    AppLogger.ui.warning("⚠️ Failed to sync routines before loading child view: \(error.localizedDescription)")
                    // Continue loading local data even if sync fails
                }
            }
            
            // Get assignments for this child
            let assignments = try await assignmentRepo.getByChild(familyId: familyId, childId: child.id)
            
            // Get routines and their steps
            var result: [RoutineWithSteps] = []
            for assignment in assignments {
                guard let routine = try await routineRepo.get(id: assignment.routineId) else { continue }
                let steps = try await stepRepo.getAll(routineId: routine.id)
                
                // Get completion state for each step
                var stepsWithCompletion: [StepWithCompletion] = []
                for step in steps {
                    let isComplete = try await deriveStepCompletion.execute(
                        familyId: familyId,
                        childId: child.id,
                        routineId: routine.id,
                        stepId: step.id,
                        localDayKey: today
                    )
                    stepsWithCompletion.append(StepWithCompletion(step: step, isComplete: isComplete))
                }
                
                result.append(RoutineWithSteps(routine: routine, steps: stepsWithCompletion))
            }
            
            routines = result
        } catch {
            self.error = error.localizedDescription
            AppLogger.error("Failed to load routines: \(error.localizedDescription)")
        }
    }
    
    func signOut() {
        do {
            try authRepository.signOut()
            AppLogger.ui.info("User signed out")
        } catch {
            AppLogger.ui.error("Error signing out: \(error.localizedDescription)")
            self.error = "Failed to sign out"
        }
    }
}

// MARK: - View Models

struct RoutineWithSteps: Identifiable {
    let routine: Routine
    let steps: [StepWithCompletion]
    
    var id: String { routine.id }
}

struct StepWithCompletion: Identifiable {
    let step: RoutineStep
    let isComplete: Bool
    
    var id: String { step.id }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/features/child/today/ChildTodayViewModel.kt
package com.HammersTech.RoutineChart.features.child.today

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.HammersTech.RoutineChart.core.data.local.SeedDataManager
import com.HammersTech.RoutineChart.core.domain.models.ChildProfile
import com.HammersTech.RoutineChart.core.domain.models.Routine
import com.HammersTech.RoutineChart.core.domain.models.RoutineStep
import com.HammersTech.RoutineChart.core.domain.repositories.AuthRepository
import com.HammersTech.RoutineChart.core.domain.repositories.ChildProfileRepository
import com.HammersTech.RoutineChart.core.domain.repositories.FamilyRepository
import com.HammersTech.RoutineChart.core.domain.repositories.RoutineAssignmentRepository
import com.HammersTech.RoutineChart.core.domain.repositories.RoutineRepository
import com.HammersTech.RoutineChart.core.domain.repositories.RoutineStepRepository
import com.HammersTech.RoutineChart.core.domain.usecases.CompleteStepUseCase
import com.HammersTech.RoutineChart.core.domain.usecases.DeriveRoutineCompletionUseCase
import com.HammersTech.RoutineChart.core.domain.usecases.DeriveStepCompletionUseCase
import com.HammersTech.RoutineChart.core.domain.usecases.UndoStepUseCase
import com.HammersTech.RoutineChart.core.utils.AppLogger
import com.HammersTech.RoutineChart.core.utils.DateHelpers
import com.HammersTech.RoutineChart.core.utils.DeviceIdentifier
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for child's today routine view
 */
@HiltViewModel
class ChildTodayViewModel @Inject constructor(
    private val seedDataManager: SeedDataManager,
    private val familyRepository: FamilyRepository,
    private val childProfileRepository: ChildProfileRepository,
    private val routineRepository: RoutineRepository,
    private val routineStepRepository: RoutineStepRepository,
    private val routineAssignmentRepository: RoutineAssignmentRepository,
    private val completeStepUseCase: CompleteStepUseCase,
    private val undoStepUseCase: UndoStepUseCase,
    private val deriveStepCompletionUseCase: DeriveStepCompletionUseCase,
    private val deriveRoutineCompletionUseCase: DeriveRoutineCompletionUseCase,
    private val deviceIdentifier: DeviceIdentifier,
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _state = MutableStateFlow(ChildTodayState())
    val state: StateFlow<ChildTodayState> = _state.asStateFlow()

    private var deviceId: String = ""

    init {
        AppLogger.UI.info("ChildTodayViewModel init started")
        initializeData()
    }

    private fun initializeData() {
        viewModelScope.launch {
            try {
                AppLogger.UI.info("Initializing data...")
                _state.update { it.copy(isLoading = true) }

                // Seed data if needed
                // Get current authenticated user ID for seeding
                val authUser = authRepository.currentUser
                if (authUser != null) {
                    AppLogger.UI.info("Seeding data if needed...")
                    seedDataManager.seedDataIfNeeded(authUser.id)
                } else {
                    AppLogger.UI.error("No authenticated user, skipping seed data")
                }

                // Get device ID
                AppLogger.UI.info("Getting device ID...")
                deviceId = deviceIdentifier.getDeviceId()
                AppLogger.UI.info("Device ID: $deviceId")

                // Load family
                AppLogger.UI.info("Loading family...")
                val family = familyRepository.getFirst()
                if (family == null) {
                    AppLogger.UI.error("No family found")
                    _state.update { it.copy(isLoading = false, error = "No family found") }
                    return@launch
                }
                AppLogger.UI.info("Family loaded: ${family.name}")

                // Load children
                AppLogger.UI.info("Loading children...")
                val children = childProfileRepository.getByFamilyId(family.id)
                if (children.isEmpty()) {
                    AppLogger.UI.error("No children found")
                    _state.update { it.copy(isLoading = false, error = "No children found") }
                    return@launch
                }
                AppLogger.UI.info("Children loaded: ${children.size}")

                _state.update {
                    it.copy(
                        family = family,
                        children = children,
                        selectedChild = children.first(),
                        isLoading = false
                    )
                }

                // Load routines for selected child
                AppLogger.UI.info("Loading routines for child...")
                loadRoutinesForChild(children.first().id, family.id, family.timeZone)
                AppLogger.UI.info("Initialization complete")
            } catch (e: Exception) {
                AppLogger.UI.error("Error initializing data: ${e.message}", e)
                android.util.Log.e("ChildTodayViewModel", "Error initializing data", e)
                _state.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun selectChild(child: ChildProfile) {
        val family = _state.value.family ?: return
        _state.update { it.copy(selectedChild = child) }
        loadRoutinesForChild(child.id, family.id, family.timeZone)
    }

    private fun loadRoutinesForChild(childId: String, familyId: String, timeZone: String) {
        viewModelScope.launch {
            try {
                val dayKey = DateHelpers.todayDayKey(timeZone)

                // Get assigned routines
                val assignments = routineAssignmentRepository.getActiveByChildId(childId)
                AppLogger.UI.info("Loading routines for childId=$childId: found ${assignments.size} assignments")
                val routineIds = assignments.map { it.routineId }
                AppLogger.UI.info("Routine IDs: ${routineIds.joinToString()}")

                val routinesWithSteps = routineIds.mapNotNull { routineId ->
                    val routine = routineRepository.getById(routineId) ?: return@mapNotNull null
                    AppLogger.UI.info("Loaded routine: ${routine.title}")
                    val steps = routineStepRepository.getByRoutineId(routineId)
                    AppLogger.UI.info("  -> ${steps.size} steps")

                    // Get completion states for all steps
                    val stepStates = deriveStepCompletionUseCase.getStepStates(
                        childId = childId,
                        routineId = routineId,
                        stepIds = steps.map { it.id },
                        dayKey = dayKey
                    )

                    // Check routine completion
                    val isComplete = deriveRoutineCompletionUseCase(
                        routine = routine,
                        childId = childId,
                        stepIds = steps.map { it.id },
                        dayKey = dayKey
                    )

                    RoutineWithSteps(
                        routine = routine,
                        steps = steps,
                        stepCompletionStates = stepStates,
                        isRoutineComplete = isComplete
                    )
                }

                _state.update { it.copy(routines = routinesWithSteps) }
            } catch (e: Exception) {
                AppLogger.UI.error("Error loading routines", e)
            }
        }
    }

    fun toggleStep(routineId: String, stepId: String) {
        val family = _state.value.family ?: return
        val child = _state.value.selectedChild ?: return

        viewModelScope.launch {
            try {
                val dayKey = DateHelpers.todayDayKey(family.timeZone)

                // Check current state
                val isCurrentlyComplete = deriveStepCompletionUseCase(
                    childId = child.id,
                    routineId = routineId,
                    stepId = stepId,
                    dayKey = dayKey
                )

                // Toggle: if complete, undo; if incomplete, complete
                if (isCurrentlyComplete) {
                    undoStepUseCase(
                        familyId = family.id,
                        familyTimeZone = family.timeZone,
                        childId = child.id,
                        routineId = routineId,
                        stepId = stepId,
                        deviceId = deviceId
                    )
                } else {
                    completeStepUseCase(
                        familyId = family.id,
                        familyTimeZone = family.timeZone,
                        childId = child.id,
                        routineId = routineId,
                        stepId = stepId,
                        deviceId = deviceId
                    )
                }

                // Reload routines to update UI
                loadRoutinesForChild(child.id, family.id, family.timeZone)
            } catch (e: Exception) {
                AppLogger.UI.error("Error toggling step", e)
            }
        }
    }

    fun signOut() {
        viewModelScope.launch {
            try {
                authRepository.signOut()
                AppLogger.UI.info("User signed out")
            } catch (e: Exception) {
                AppLogger.UI.error("Error signing out: ${e.message}", e)
                _state.update { it.copy(error = "Failed to sign out") }
            }
        }
    }
}

data class ChildTodayState(
    val isLoading: Boolean = true,
    val error: String? = null,
    val family: com.HammersTech.RoutineChart.core.domain.models.Family? = null,
    val children: List<ChildProfile> = emptyList(),
    val selectedChild: ChildProfile? = null,
    val routines: List<RoutineWithSteps> = emptyList()
)

data class RoutineWithSteps(
    val routine: Routine,
    val steps: List<RoutineStep>,
    val stepCompletionStates: Map<String, Boolean>,
    val isRoutineComplete: Boolean
)

